modules/_labs/tabs/react/README.md

```diff
@@ -0,0 +1,49 @@
+# Canvas Kit React Tabs
+
+<a href="https://github.com/Workday/canvas-kit/tree/master/modules/_labs/README.md">
+  <img src="https://img.shields.io/badge/LABS-alpha-orange" alt="LABS: Alpha" />
+</a>  This component is work in progress and currently in pre-release.
+
+Composable tab components
+
+## Installation
+
+```sh
+yarn add @workday/canvas-kit-labs-react-tabs
+```
+
+## Usage
+
+```tsx
+import * as React from 'react';
+import Tabs, {TabList, Tab, TabPanels, TabPanel} from '@workday/canvas-kit-labs-react-tabs';
+
+  <Tabs>
+    <TabList>
+      <Tab index={0}>Tab</Tab>
+      <Tab index={1}>Tab Number Two</Tab>
+    </TabList>
+    <TabPanels>
+      <TabPanel key="1" index={0}>
+        Hi this is tab content
+      </TabPanel>
+      <TabPanel key="2" index={1}>
+        Hi this is another tab's content
+      </TabPanel>
+    </TabPanels>
```

Comment:
I've thought a lot about tabs including the API and performance. I know Reach UI does this same API: https://reacttraining.com/reach-ui/tabs/

My concern with a `TabPanel` having all the VDOM present of every render is the potential for more expensive React renders even if non-active tabs aren't rendered to the DOM. I noticed the implementation currently renders all tab panels to the DOM and uses `hidden`. This could be a very big initial render hit on tabs for large tab panel contents.

I think iterating over `TabPanel` children can short-circuit child render functions (React does top-down rendering), but the JSON object still has to be created. Have you run into any problems with this? Also the tab content could be huge. Did you solve this by having a component per `TabPanel` content? Was there any performance concerns with iterating over children using `React.Children.map` and `React.cloneElement`?

The API is definitely easier to understand by looking at the JSX. While the JSX makes it easy to understand TabPanel content, there can be other issues understanding how content gets projected. Is the `TabPanel` that shows up in the final React render tree still a `TabPanel`? Do extra props get mixed in using `React.cloneElement`? The interface of `TabPanel` would then have to include optional props that the application developer shouldn't really be exposed to. If a different component is rendered, it gets a bit confusing how the source JSX translates to the final JSX.

I don't think there is a perfect API. Some components have child lists that could be potentially huge and we have more of a renderProp API for those instances. The API is a bit harder to understand, but does allow us to add rendering optimizations with less confusion of the final JSX and DOM.

Reply:
> This could be a very big initial render hit on tabs for large tab panel contents.

At least for how we used the tabs in our app, we had performance issues with the tabs being very laggy when clicking between them and causing really choppy interactions. We studied a couple different tab components and it seemed like rendering the panels up front as `hidden` was the fix to this problem. We were rending large lists with images and many svg icons which made the render of each panel slow. 

Surely we could have added virtual scrolling etc to make the individual panels quicker if we want to speed up the initial render. I think it may be a matter of preference if we would like to trade the speed of the initial render for the speed of renders on selecting tabs. Perhaps we prefer to speed up the first render and to force our consumers tightly manage the performance of their panels.

> Is the `TabPanel` that shows up in the final React render tree still a `TabPanel`?

Right now, the `TabPanel` components written by the user will end up in the dom as a div with the role `tabpanel`. Is this too far away from what users might associate with the `TabPanel` component itself?

> The interface of `TabPanel` would then have to include optional props that the application developer shouldn't really be exposed to. 

What kind of props do you imagine we might need on the `TabPanel`? I will need to add the accessibility hookups but I think those can be done without exposing any props on the component's API.

> Was there any performance concerns with iterating over children using React.Children.map and React.cloneElement?, Do extra props get mixed in using React.cloneElement?

I also don't see any need to cloneElement right now on the tab panels if we want to force the indexes to be required. If the indexes are optional for both the tabs and the tab panels, we will have to clone element to get the lists built with proper indexes. I think I prefer the API of optional indexes, but like you said, the panel content could be large and could slow down performance.

Reply:
> This could be a very big initial render hit on tabs for large tab panel contents.

Rendering tab panel content with a `hidden` attribute will be more efficient when the browser does Layout and Paint, but won't be any faster inside React. ReactDOM will still process VDOM and turn it into DOM and will still run reconciliation on `TabPanel` children on changes.

What I mean here is APIs like making the contents of `TabPanel` be a function (render prop) or `null`ing the children inside the `Tab` component when iterating over `TabPanels`. This would still parse VDOM objects, but prevent the ReactDOM reconciler from processing and creating DOM of `TabPanel` content that is not active.

> Is the `TabPanel` that shows up in the final React render tree still a `TabPanel`?

I confirmed:
<img width="156" alt="Screen Shot 2020-04-22 at 11 10 09 AM" src="https://user-images.githubusercontent.com/338257/80012552-b7123b00-848a-11ea-93ec-d4e7a6270d13.png">

I meant the React VDOM tree and not the DOM tree. That screenshot is of the React Component render tree using the React dev tools.

> The interface of `TabPanel` would then have to include optional props that the application developer shouldn't really be exposed to.

The accessibility attributes are what I meant. They have to be valid and optional on the `TabPanel` prop interface even if the developer never sets them. They will be overridden I'm sure. You could avoid this by modifying the `React.cloneElement` call and saying what extra props are allowed. This way the external interface won't mention these attributes are valid to set.

> Was there any performance concerns with iterating over children using React.Children.map and React.cloneElement?, Do extra props get mixed in using React.cloneElement?

I think this question covers both performance of rendering DOM (even if using `hidden`) and having to add ARIA attributes. We may be using `cloneElement` right now only to add `key`, but we'll use it to add other attributes as well.